File: src/10-word-count/c/main.c
/*!
 * Polyglot Programming
 *
 * Copyright (c) 2024 Adam Burucs
 *
 * MIT Licensed
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int countWords(const char *fileName, int *wordCount) {
    FILE *file = fopen(fileName, "r");
    if (file == NULL) {
        return -1; // Error opening file
    }

    char line[1024];
    *wordCount = 0;

    while (fgets(line, sizeof(line), file)) {
        // Split the line into words
        char *token = strtok(line, " \t\n\r");
        while (token != NULL) {
            (*wordCount)++;
            token = strtok(NULL, " \t\n\r");
        }
    }

    fclose(file);
    return 0; // Success
}

int main() {
    printf("Count total words sum\n\n");

    int wordCount;
    int result = countWords("enchiridion-hun.txt", &wordCount);

    if (result == 0) {
        printf("Total words: %d\n", wordCount);
    } else {
        printf("Error: Could not open file.\n");
    }

    return 0;
}

File: src/10-word-count/typescript/src/main.ts
/*!
 * Polyglot Programming
 *
 * Copyright (c) 2024- Adam Burucs
 *
 * MIT Licensed
 */

import fs from "node:fs/promises";

async function countWords(fileName: string): Promise<number | string> {
  const file = await fs.open(fileName);
  let count = 0;

  for await (const line of file.readLines()) {
    let splitLine = line.split(" ");
    count += splitLine.length;
  }

  return count;
}

async function main(): Promise<void> {
  console.log("Count total words sum");
  console.log();

  try {
    let count = await countWords("enchiridion-hun.txt");
    if (typeof count === "number") {
      console.log(`Total words: ${count}`);
    } else if (typeof count === "string") {
      console.log(`Error: ${count}`);
    }
  } catch (error) {
    console.error("An error occurred:", error);
  }
}

main();
